compiler/circle-mpqsolver/src/bisection/ErrorApproximator.cpp

```diff
@@ -0,0 +1,386 @@
+    {
+      for (indices[2] = 0; indices[2] < dimension.dim(2).value(); indices[2]++)
+      {
+        for (indices[3] = 0; indices[3] < dimension.dim(3).value(); indices[3]++)
+        {
+          func(indices, dimension, channel_dim_index);
+        }
+      }
+    }
+  }
+}
+
+void cal_minmax_per_channel(const CircleNode *node, std::vector<float> &min,
+                            std::vector<float> &max)
+{
+  CircleConst *weight = loco::must_cast<CircleConst *>(get_weight(node));
+  loco::TensorShape dimension;
+  set_weight_dim(node, weight, dimension);
+
+  auto channel_dim_index = get_channel_dim_index(node);
+  auto size = dimension.dim(channel_dim_index).value();
+
+  std::vector<bool> has_min_max_value(size, false);
+  min.resize(size);
+  max.resize(size);
+
+  auto cal_minmax = [&](uint32_t *indices, loco::TensorShape &dimension,
+                        uint32_t channel_dim_index) {
+    uint32_t channel_idx = indices[channel_dim_index];
+    auto data = weight->at<loco::DataType::FLOAT32>(cal_offset(dimension, indices));
+    if (has_min_max_value[channel_idx])
+    {
+      min[channel_idx] = data < min[channel_idx] ? data : min[channel_idx];
+      max[channel_idx] = data > max[channel_idx] ? data : max[channel_idx];
+    }
+    else
+    {
+      min[channel_idx] = data;
+      max[channel_idx] = data;
+      has_min_max_value[channel_idx] = true;
+    }
+  };
+
+  iterate_per_channel(node, cal_minmax);
+}
+
+bool get_shape(const CircleNode *circle_node, std::vector<uint32_t> &shape)
+{
+  if (circle_node->shape_status() == ShapeStatus::VALID)
+  {
+    auto rank = circle_node->rank();
+    if (rank != 4)
+      return false;
+
+    shape.resize(rank);
+    for (uint32_t i = 0; i < rank; i++)
+    {
+      shape[i] = circle_node->dim(i).value();
+    }
+    return true;
+  }
+
+  return false;
+}
+
+const CircleNode *get_activation_node(const CircleNode *node)
+{
+  // only convolutions are considered currently, so just one input
+  auto preds = loco::preds(node);
+  for (const auto &pred : preds)
+  {
+    auto cur_node = loco::must_cast<const CircleNode *>(pred);
+    if (cur_node->opcode() != CircleOpcode::CIRCLECONST)
+    {
+      return cur_node;
+    }
+  }
+  return nullptr;
+}
+
+uint32_t get_stride_multiplier(const CircleNode *node)
+{
+  uint32_t mult = 1;
+  if (node->opcode() == CircleOpcode::DEPTHWISE_CONV_2D)
+  {
+    auto dconv = loco::must_cast<const CircleDepthwiseConv2D *>(node);
+    auto stride = dconv->stride();
+    mult *= (stride->h() * stride->w());
+  }
+  else if (node->opcode() == CircleOpcode::CONV_2D)
+  {
+    auto nconv = loco::must_cast<const CircleConv2D *>(node);
+    auto stride = nconv->stride();
+    mult *= (stride->h() * stride->w());
+  }
+  else if (node->opcode() == CircleOpcode::TRANSPOSE_CONV)
+  {
+    auto nconv = loco::must_cast<const CircleTransposeConv *>(node);
+    auto stride = nconv->stride();
+    mult *= (stride->h() * stride->w());
+  }
```

Comment:
Please use switch statement and throw an exception in default case. This is to avoid potential bug..

Reply:
ok